### ISSUE TITLE
Custom Dataset: Number of Classes

### ISSUE THREAD

**Issue by ad12** [bonlime/keras-deeplab-v3-plus#26]
Should `num_classes` take into account the background class. For example, if I have 3 foreground classes that I have masks for and a background class that I don't care about and don't have a mask for, should my `num_classes` be 3?

**Comment by bonlime** [comment]
It depends on your loss function and classes topology.  
If classes are mutually exclusive you can add background class and use softmax at the top. During training at Imagenet authors of original paper used this way.  
But you can use sigmoid as activation at the last layer without adding background and it also will work

**Comment by vodp** [comment]
As @bonlime said if you background contains other classes you may not want to assign it as a 'background' class, rather just ignore it as 'ignore_label'. Of course in detection the network segments every pixel so that you won't have background being classified.

If background is really irrelevant objects, then you may want to add it as an additional class.

In VOC pascal the Deeplabv3 used background class as it was defined in the dataset.
In city datasets, background class is absent because dense segmentation was designed for those datasets.

**Comment by ad12** [comment]
Thank you!

**Comment by ad12** [comment]
@vodp I am planning on not adding a background class, but how do I ignore the pixel? Is there a way to specify it as ignore_label?

I am also not getting a binary mask after training. Is there any changes that I need to make for this?

**Comment by vodp** [comment]
You need to write a custom loss function, addressing ignore_label and do not include the loss incurred by ignore_label into the total loss.
